sharktank/tests/models/llama4/moe_test.py

```diff
@@ -0,0 +1,220 @@
+                theta=shared_ffn_theta,
+                activation_fn=torch.nn.functional.silu,
+            )
+
+        hf_moe = Llama4TextMoe(
+            num_experts_per_tok=expert_used_count,
+            hidden_size=theta.tensor("ffn_gate_exps", "weight").shape[2],
+            num_local_experts=theta.tensor("ffn_gate_exps", "weight").shape[0],
+            intermediate_size=theta.tensor("ffn_gate_exps", "weight").shape[1],
+            act_fn=torch.nn.functional.silu,
+            shared_expert=shared_experts,
+        )
+        hf_moe.router.weight.data = unbox_tensor(theta("ffn_gate_inp.weight"))
+        hf_moe.experts.gate_up_proj.data = torch.cat(
+            [
+                unbox_tensor(theta("ffn_gate_exps.weight")),
+                unbox_tensor(theta("ffn_up_exps.weight")),
+            ],
+            dim=1,
+        ).transpose(1, 2)
+        hf_moe.experts.down_proj.data = unbox_tensor(
+            theta("ffn_down_exps.weight")
+        ).transpose(1, 2)
+
+        hf_moe.shared_expert.gate_proj.weight.data = unbox_tensor(
+            theta("ffn_gate_shexp.weight")
+        )
+        hf_moe.shared_expert.up_proj.weight.data = unbox_tensor(
+            theta("ffn_up_shexp.weight")
+        )
+        hf_moe.shared_expert.down_proj.weight.data = unbox_tensor(
+            theta("ffn_down_shexp.weight")
+        )
+
+        hf_res = hf_moe(input)[0]
+        hf_moe_output = hf_res.reshape(input.shape)
+        torch.testing.assert_close(moe_output, hf_moe_output, atol=1e-3, rtol=1e-2)
+
+
+class Llama4TextMoe(torch.nn.Module):
+    def __init__(
+        self,
+        num_experts_per_tok: int,
+        hidden_size: int,
+        num_local_experts: int,
+        intermediate_size: int,
+        act_fn,
+        shared_expert: FFN,
+    ):
+        super().__init__()
+        self.top_k = num_experts_per_tok
+        self.hidden_dim = hidden_size
+        self.num_experts = num_local_experts
+        self.experts = Llama4TextExperts(
+            num_local_experts, intermediate_size, hidden_size
+        )
+        self.router = torch.nn.Linear(hidden_size, num_local_experts, bias=False)
+        self.shared_expert = Llama4TextMLP(hidden_size, intermediate_size)
+        # self.shared_expert = shared_expert
+
+    def forward(self, hidden_states):
+        batch, seq_len, hidden_dim = hidden_states.shape
+        hidden_states = hidden_states.view(-1, self.hidden_dim)
+        router_logits = self.router(hidden_states).transpose(0, 1)
+        tokens_per_expert = batch * seq_len
+
+        router_top_value, router_indices = torch.topk(
+            router_logits.transpose(0, 1), self.top_k, dim=1
+        )
+        router_scores = (
+            torch.full_like(router_logits.transpose(0, 1), float("-inf"))
+            .scatter_(1, router_indices, router_top_value)
+            .transpose(0, 1)
+        )
+        # We do this to make sure we have -inf for non topK tokens before going through the !
```

Comment:
"before going through the !" ?

Reply:
The latest Transformers version does not contain these comments. I have updated the code file.